src/bokeh/core/property/descriptors.py

```diff
@@ -756,6 +758,9 @@ def __set__(self, obj, value, *, setter=None):
         from ...document.events import ColumnDataChangedEvent
         hint = ColumnDataChangedEvent(obj.document, obj, "data", setter=setter) if obj.document else None
 
+        # Try converting to Narwhals so that any dataframe supported by Narwhals
+        # can be supported.
+        value = nw.from_native(value, eager_only=True, allow_series=True, pass_through=True)
```

Comment:
coudl this not also go into `prepare_value`? That seems like the right place. I'll have to think on it more but I feel like there should not be narwhals-specific code in `descriptors.py`